qiskit/dagcircuit/dagcanonical.py

```diff
@@ -0,0 +1,400 @@
+'successors': [], 'predecessors': [1]})]
+
+The reference paper is https://arxiv.org/abs/1909.05270v1
+
+"""
+
+from collections import OrderedDict
+import copy
+import heapq
+import networkx as nx
+import numpy as np
+
+from qiskit.circuit.quantumregister import QuantumRegister
+from qiskit.circuit.classicalregister import ClassicalRegister
+from qiskit.quantum_info.operators import Operator
+from .exceptions import DAGCircuitError
+from .dagnode import DAGNode
+
+
+class DAGcanonical:
+    """
+    Object to represent a quantum circuit as a DAG in the canonical form.
+    """
+    def __init__(self):
+        """
+        Create an empty directed acyclis graph (canonical form)
+        """
+        # Circuit name
+        self.name = None
+
+        # Directed multigraph whose nodes are operations(gates) and edges
+        # represent non-commutativity between two gates.
+        self._multi_graph = nx.MultiDiGraph()
+
+        # Map of qreg name to QuantumRegister object
+        self.qregs = OrderedDict()
+
+        # Map of creg name to ClassicalRegister object
+        self.cregs = OrderedDict()
+        # Index of the last node added
+        self._max_node_id = 0
+
+        # Intern list of nodes
+        self._id_to_node = {}
+
+    def to_networkx(self):
+        """Returns a copy of the DAGCircuit in networkx format."""
+        return copy.deepcopy(self._multi_graph)
+
+    def qubits(self):
+        """Return a list of qubits (as a list of Qubit instances)."""
+        return [qubit for qreg in self.qregs.values() for qubit in qreg]
+
+    def node_counter(self):
+        """ Returns the number of nodes in the dag """
+        return len(self._multi_graph)
+
+    def add_qreg(self, qreg):
+        """Add qubits in a quantum register."""
+        if not isinstance(qreg, QuantumRegister):
+            raise DAGCircuitError("not a QuantumRegister instance.")
+        if qreg.name in self.qregs:
+            raise DAGCircuitError("duplicate register %s" % qreg.name)
+        self.qregs[qreg.name] = qreg
+
+    def add_creg(self, creg):
+        """Add all wires in a classical register."""
+        if not isinstance(creg, ClassicalRegister):
+            raise DAGCircuitError("not a ClassicalRegister instance.")
+        if creg.name in self.cregs:
+            raise DAGCircuitError("duplicate register %s" % creg.name)
+        self.cregs[creg.name] = creg
+
+    def add_node(self, operation, qargs, cargs):
+        """Add a DAGnode to the graph.
+
+        Args:
+            operation (Instruction): operation as a quantum gate.
+            qargs (list[Qubit]): list of qubits on which the operation acts
+            cargs (list[Clbit]): list of classical wires to attach to.
+        """
+        node_properties = {
+            "type": "op",
+            "op": operation,
+            "name": operation.name,
+            "qargs": qargs,
+            "cargs": cargs,
+            "condition": operation.condition
```

Comment:
Could you use `qiskit/dagcircuit/dagnode.py` here, instead of using python dictionaries for the nodes?